src/libraries/System.Private.CoreLib/src/System/Number.DecimalIeee754.cs

```diff
@@ -0,0 +1,255 @@
+        static abstract TValue G0ToGwPlus1ExponentMask { get; } //G0 to G(w+1)
+        static abstract TValue G2ToGwPlus3ExponentMask { get; } //G2 to G(w+3)
+        static abstract TValue GwPlus2ToGwPlus4SignificandMask { get; } //G(w+2) to G(w+4)
+        static abstract TValue GwPlus4SignificandMask { get; } //G(w+4)
+        static abstract int NumberBitsSignificand { get; }
+        static abstract int NumberDigitsPrecision { get; }
+        static abstract int Bias { get; }
+        static abstract TValue MostSignificantBitOfSignificandMask { get; }
+        static abstract int ConvertToExponent(TValue value);
+        static abstract TSignificand ConvertToSignificand(TValue value);
+        static abstract TSignificand Power10(int exponent);
+    }
+
+    internal static partial class Number
+    {
+        internal static TValue CalDecimalIeee754<TDecimal, TSignificand, TValue>(TSignificand significand, int exponent)
+            where TDecimal : unmanaged, IDecimalIeee754ConstructorInfo<TDecimal, TSignificand, TValue>
+            where TSignificand : IBinaryInteger<TSignificand>
+            where TValue : IBinaryInteger<TValue>
+        {
+            if (significand == TSignificand.Zero)
+            {
+                return TValue.Zero;
+            }
+
+            TSignificand unsignedSignificand = significand > TSignificand.Zero ? significand : -significand;
+
+            if ((unsignedSignificand > TDecimal.MaxSignificand && exponent >= TDecimal.MaxDecimalExponent)
+                || (unsignedSignificand == TDecimal.MaxSignificand && exponent > TDecimal.MaxDecimalExponent))
+            {
+                return significand > TSignificand.Zero ? TDecimal.PositiveInfinityBits : TDecimal.NegativeInfinityBits;
+            }
+
+            TSignificand ten = TSignificand.CreateTruncating(10);
+            if (exponent < TDecimal.MinDecimalExponent)
+            {
+                while (unsignedSignificand > TSignificand.Zero && exponent < TDecimal.MinDecimalExponent)
+                {
+                    unsignedSignificand /= ten;
+                    ++exponent;
+                }
+                if (unsignedSignificand == TSignificand.Zero)
+                {
+                    return TDecimal.Zero;
+                }
+            }
+
+            if (unsignedSignificand > TDecimal.MaxSignificand)
+            {
+                int numberDigitsRemoving = TDecimal.CountDigits(unsignedSignificand) - TDecimal.NumberDigitsPrecision;
+
+                if (exponent + numberDigitsRemoving > TDecimal.MaxDecimalExponent)
+                {
+                    throw new OverflowException(TDecimal.OverflowMessage);
+                }
+
+                exponent += numberDigitsRemoving;
+                TSignificand divisor = TDecimal.Power10(numberDigitsRemoving);
+                TSignificand quotient = unsignedSignificand / divisor;
+                TSignificand remainder = unsignedSignificand % divisor;
```

Comment:
Another bit of code where there is an existing helper that can handle various optimizations:
```suggestion
                (TSignificand quotient, TSignificand remainder) = TSignificand.DivRem(unsignedSignificand, divisor);
```